src/watchmaker/workers/salt.py

```diff
@@ -360,6 +381,97 @@ def _get_failed_states(self, state_ret):
             failed_states = state_ret
         return failed_states
 
+    def _install_pip(self):
+        if not self.py_exec:
+            raise WatchmakerException("Python executable not set")
+
+        get_pip = os.path.join(
+            os.path.abspath(
+                os.path.join(os.path.dirname(__file__), '..', '..')),
+            'vendor',
+            'pypa',
+            'get-pip',
+            '2.6',
+            'get-pip.py')
+        self.log.info(
+            'Attempting pip install using get-pip (%s)...', get_pip)
+        cmd = [
+            self.py_exec,
+            get_pip,
+            "--index-url",
+            self.pip_index
+        ]
+        self.call_process(cmd)
+
+    def _upgrade_pip(self):
+        if not self.py_exec:
+            raise WatchmakerException("Python executable not set")
+
+        self.log.info('Attempting to upgrade pip...')
+        cmd = [
+            self.py_exec,
+            "-m",
+            "pip",
+            "install",
+            "--upgrade",
+            "pip",
+            "--index-url",
+            self.pip_index
+        ]
+        try:
+            self.call_process(cmd)
+        except WatchmakerException:
+            self.log.info(
+                'Unable to upgrade pip, continuing with legacy version')
+
+        ver = self.call_process([self.py_exec, '-m', 'pip', '--version'])
+        self.log.info('Pip version: %s', ver['stdout'])
+
+    def _install_pip_packages(self):
+        if not self.pip_install:
+            self.log.info('No Python packages specified for install in config')
+            return 0
+
+        self.py_exec = self._get_grain('pythonexecutable')
+        self.log.info('Salt Python interpreter: `%s`', self.py_exec)
+
+        try:
+            ver = self.call_process([
+                self.py_exec,
+                '-m',
+                'pip',
+                '--version'])
+            self.log.info('Pip version: %s', ver['stdout'])
+        except WatchmakerException:
+            self.log.info('Pip not installed for Salt interpreter!')
+            self._install_pip()
+
+        self._upgrade_pip()
+
+        self.log.info(
```

Comment:
This one makes sense as `info`, since it is directly related to something the user asked us to do....